Code formatting

dNechita opened the pull request:
## PR Description
This PR continues the effort to introduce code formatting rules, initially discussed in #397 and further developed in #413 by @rgetz.
Additionally, this PR provides a script that enables users to format the entire codebase.

## PR Type
- [ ] Bug fix (a change that fixes an issue)
- [x] New feature (a change that adds new functionality)
- [ ] Breaking change (a change that affects other repos or cause CIs to fail)

## PR Checklist
- [x] I have conducted a self-review of my own code changes
- [x] I have commented new code, particularly complex or unclear areas
- [x] I have checked that I did not introduce new warnings or errors (CI output)
- [ ] I have checked that components that use libiio did not get broken
- [ ] I have updated the documentation accordingly (GitHub Pages, READMEs, etc)

rgetz committed:
add default formatting templates for clang-format and cmake-format

per discussion in #397, The clang-format is the linux kernel code formatting
template with a few minor tweaks:

indent continuation more:

-ContinuationIndentWidth: 8
+ContinuationIndentWidth: 16

remove the kernel's ForEachMacros

Sort the include paths, like Lars suggested in #411

-SortIncludes: false
+SortIncludes: true

In this order:

-#IncludeBlocks: Preserve # Unknown to clang-format-5.0
+IncludeBlocks: Regroup
 IncludeCategories:
-  - Regex: '.*'
-    Priority: 1
-IncludeIsMainRegex: '(Test)?$'
+  - Regex:           '^<.*\.h>'
+    Priority:        1
+  - Regex:           '^<.*'
+    Priority:        2
+  - Regex:           '.*'
+    Priority:        3
+IncludeIsMainRegex: 'xxx'

The cmake-format.py file is from the doc, with tabs set to 4 spaces
https://cmake-format.readthedocs.io/en/latest/configuration.html

The cmake-format.py has been configured to not touch the comments
because some formatting isn't done right. For example, a comment
containing a function and its parameters got formatted and lots
of spaces were introduces making everything unaligned. The comments
format shouldn't be that strict.

This is something for discussion, This is expected to have
clang-format version 7.0.1-8 (tags/RELEASE_701/final) or higher.

Signed-off-by: Robin Getz <robin.getz@analog.com>

rgetz committed:
general tweaks for code updates - no functional changes

Signed-off-by: Robin Getz <robin.getz@analog.com>

dNechita committed:
code-format: Add script that formats entire code base

The script supports ignoring specific files and directories by reading
their paths from a list, ensuring that any files or folders specified
will be excluded from formatting.

Signed-off-by: Dan Nechita <dan.nechita@analog.com>

mhennerich commented on .clang-format:
```diff
@@ -0,0 +1,134 @@
+BreakConstructorInitializers: BeforeComma
+BreakAfterJavaFieldAnnotations: false
+BreakStringLiterals: false
+ColumnLimit: 80
```
Looking through the diff. There are lots of excessive line breaks!
I think we should go for al least 100.
https://lkml.org/lkml/2020/5/29/1317

rgetz replied:
At one time - I made a loop, that incremented by 5's and then saved off where the min diff was - I don't remember what the results were - but it would be easy to do it again...

dNechita committed:
code-format: Don't let formatter alter the include order for dns files

On Windows, the include order matters so turn off clang-format to
preserve the include order.
Otherwise the windows build will fail.

Signed-off-by: Dan Nechita <dan.nechita@analog.com>

dNechita commented:
V1:

- Extended the maximum line length from 80 to 100
- Prevented clang-format from reordering the includes on some files which was causing windows builds to fail (dns_sd.h, dns_sd_windows.c)
- Fixed 2 spelling errors in cmake-format.py

rgetz commented:
After fixing things - a pre-commit hook/check is needed to ensure that all future patches keep the style - or it will just start diverting...

dNechita commented:
> After fixing things - a pre-commit hook/check is needed to ensure that all future patches keep the style - or it will just start diverting...

I have an active PR (#1338) that implements a CI job that checks if code is formatted. The job will fail if it's not. Specifically, the job checks all changed files, applies formatting, and fails if even one is not properly formatted.
The job is triggered on pull requests.

999Marius reviewed (approved):
LGTM

dNechita commented:
V2:

- Rebased on latest main
- Added  "zephyr" dir to the ignore list for both clang-format and cmake-format
- Configured cmake-format to not touch comments as I messes up some of the comments
- Reapplied the formatting on the entire codebase

dNechita committed:
clang-format: Don't align backslashes

Some macros (e.g., IIO_ENSURE, error) are not formatted correctly due to a
mix of tabs and spaces. The alignment of backslashes (positioned to match
the longest line) depends on how each editor renders tab width.

Set 'AlignEscapedNewlines' to 'DontAlign' to solve this.

Signed-off-by: Dan Nechita <dan.nechita@analog.com>

dNechita committed:
treewide: Format entire codebase

Format all .c and .h files using rules from .clang-format but ignore
paths defined in .clangformatignore file.

Format all .cmake and CMakeLists.txt files using rules from
cmake-format.py but ignore paths defined in .cmakeformatignore file.

The current changes result from executing:
./format.sh
in the repository's top-level directory.

Signed-off-by: Dan Nechita <dan.nechita@analog.com>

dNechita commented:
V3:

- Don't align backslashes. This way we avoid this kind of formatting:
<img width="1562" height="679" alt="image" src="https://github.com/user-attachments/assets/65b611ad-b1cc-4d70-933b-72e14e8a4eb7" />
